### DISCUSSION TITLE
Robust Scheduling

### DISCUSSION THREAD

**Discussion by joeb8799** [NatLabRockies/ROSCO#149] [General]
Hello,
after generating the linearization files for the IEA15MW turbine, I'm trying to run the robust scheduling method in order to get the values of omega for every wind speed in the U_pc given in the yaml input file.
I got the following error:

/home/g69581/anaconda3/envs/rosco-env/lib/python3.8/site-packages/openmdao/approximation_schemes/finite_difference.py:135: OMDeprecationWarning:When using 'rel' as the step_calc, the fd stepsize is currently scaled by the norm of the vector variable. This is not ideal for larger vectors, and this behavior is being changed in OpenMDAO 3.12.0. To preserve the older way of doing this calculation, set step_calc to 'rel_legacy'.
/home/g69581/anaconda3/envs/rosco-env/lib/python3.8/site-packages/openmdao/core/total_jac.py:1568: DerivativesWarning:Constraints or objectives [('r_sched.sm', inds=[0])] cannot be impacted by the design variables of the problem.
Optimization terminated successfully    (Exit mode 0)
            Current function value: -0.3
            Iterations: 2
            Function evaluations: 2
            Gradient evaluations: 2
Optimization Complete

and after that error, i got the same value of omega ( 0.3) for every wind speed in U_pc.
How can i fix it?

**Comment by nikhar-abbas** [comment]
Hi,

Happy to hear that you are using the robust scheduling features. 

First, it looks like OpenMDAO changed the API and a minor change to ROSCO toolbox source code is necessary to accommodate this.  We will try to get that sorted in the coming weeks. For now, I would just change [this line ](https://github.com/NREL/ROSCO/blob/77278f5d1cee95a830b22d79797b351481e87972/ROSCO_toolbox/linear/robust_scheduling.py#L275) to have `step_calc=rel_avg` for now, and hopefully this mitigates this error.

Second, which branch of ROSCO are you on? If you are not on develop, I would start with checking out this branch and seeing if these error messages persist. Hopefully some of the updates made in #139 help fix your issues.

Finally, I am curious about your problem setup. Are you closely mirroring [Example 12](https://github.com/NREL/ROSCO/blob/develop/Examples/example_12.py)? If not, I would certainly start there. 

Cheers,
Nikhar

**Reply by joeb8799** [reply]
Hello Mr.Abbas,
Thank you so much for your answer.
I followed all your instructions, changed step_calc=rel_avg and then used the develop branch of the ROSCO toolbox and checked if all the updates are present in the files.
But I still have the same problem. I want to clarify that I am imitating example 12 to get the omega values for the IEA-15-240-RWT-UMaineSemi.
The error I get is as follows:
loaded in parallel in 5.5002357959747314 seconds
Finding ROSCO tuning parameters for u = 12, sm = 0.1, omega_pc = [0.0, 0.3], k_float = 0.0
omega = 0.1, sm = 0.999999999992498
omega = 0.1, sm = 0.999999999992498
omega = 0.11000000000000001, sm = 0.999999999992498
omega = 0.09, sm = 0.999999999992498
omega = 0.11000000000000001, sm = 0.999999999992498
omega = 0.09, sm = 0.999999999992498
/home/g69581/anaconda3/envs/rosco-env/lib/python3.8/site-packages/openmdao/core/total_jac.py:1568: DerivativesWarning:Constraints or objectives [('r_sched.sm', inds=[0])] cannot be impacted by the design variables of the problem.
omega = 0.3, sm = 0.999999999992498
omega = 0.31, sm = 0.999999999992498
omega = 0.29, sm = 0.999999999992498
Optimization terminated successfully    (Exit mode 0)
            Current function value: -0.3
            Iterations: 2
            Function evaluations: 2
            Gradient evaluations: 2
Optimization Complete
-----------------------------------
Finding ROSCO tuning parameters for u = 14, sm = 0.1, omega_pc = [0.0, 0.3], k_float = 0.0
omega = 0.1, sm = 0.9999999999970182
omega = 0.1, sm = 0.9999999999970182
omega = 0.11000000000000001, sm = 0.9999999999970182
omega = 0.09, sm = 0.9999999999970182
/home/g69581/anaconda3/envs/rosco-env/lib/python3.8/site-packages/openmdao/core/total_jac.py:1568: DerivativesWarning:Constraints or objectives [('r_sched.sm', inds=[0])] cannot be impacted by the design variables of the problem.
omega = 0.3, sm = 0.9999999999970182
omega = 0.31, sm = 0.9999999999970182
omega = 0.29, sm = 0.9999999999970182
Optimization terminated successfully    (Exit mode 0)
            Current function value: -0.3
            Iterations: 2
            Function evaluations: 2
            Gradient evaluations: 2
Optimization Complete
-----------------------------------
Finding ROSCO tuning parameters for u = 15, sm = 0.1, omega_pc = [0.0, 0.3], k_float = 0.0
omega = 0.1, sm = 0.9999999999954957
omega = 0.1, sm = 0.9999999999954957
omega = 0.11000000000000001, sm = 0.9999999999954957
omega = 0.09, sm = 0.9999999999954957
/home/g69581/anaconda3/envs/rosco-env/lib/python3.8/site-packages/openmdao/core/total_jac.py:1568: DerivativesWarning:Constraints or objectives [('r_sched.sm', inds=[0])] cannot be impacted by the design variables of the problem.
omega = 0.3, sm = 0.9999999999954957
omega = 0.31, sm = 0.9999999999954957
omega = 0.29, sm = 0.9999999999954957
Optimization terminated successfully    (Exit mode 0)
            Current function value: -0.3
            Iterations: 2
            Function evaluations: 2
            Gradient evaluations: 2
Optimization Complete
-----------------------------------
Finding ROSCO tuning parameters for u = 18, sm = 0.1, omega_pc = [0.0, 0.3], k_float = 0.0
omega = 0.1, sm = 0.9999999999954617
omega = 0.1, sm = 0.9999999999954617
omega = 0.11000000000000001, sm = 0.9999999999954617
omega = 0.09, sm = 0.9999999999954617
/home/g69581/anaconda3/envs/rosco-env/lib/python3.8/site-packages/openmdao/core/total_jac.py:1568: DerivativesWarning:Constraints or objectives [('r_sched.sm', inds=[0])] cannot be impacted by the design variables of the problem.
omega = 0.3, sm = 0.9999999999954617
omega = 0.31, sm = 0.9999999999954617
omega = 0.29, sm = 0.9999999999954617
Optimization terminated successfully    (Exit mode 0)
            Current function value: -0.3
            Iterations: 2
            Function evaluations: 2
            Gradient evaluations: 2
Optimization Complete
-----------------------------------
Finding ROSCO tuning parameters for u = 20, sm = 0.1, omega_pc = [0.0, 0.3], k_float = 0.0
omega = 0.1, sm = 0.999999999994936
omega = 0.1, sm = 0.999999999994936
omega = 0.11000000000000001, sm = 0.999999999994936
omega = 0.09, sm = 0.999999999994936
/home/g69581/anaconda3/envs/rosco-env/lib/python3.8/site-packages/openmdao/core/total_jac.py:1568: DerivativesWarning:Constraints or objectives [('r_sched.sm', inds=[0])] cannot be impacted by the design variables of the problem.
omega = 0.3, sm = 0.999999999994936
omega = 0.31, sm = 0.999999999994936
omega = 0.29, sm = 0.999999999994936
Optimization terminated successfully    (Exit mode 0)
            Current function value: -0.3
            Iterations: 2
            Function evaluations: 2
            Gradient evaluations: 2
Optimization Complete
-----------------------------------
[0.3, 0.3, 0.3, 0.3, 0.3] [1.2 1.2 1.2 1.2 1.2]

Thank you in advance and Have a good day.
Regards.

**Comment by nikhar-abbas** [comment]
This is indeed strange behavior - are you using the linearized IEA15MW files provided here: https://github.com/NREL/ROSCO/tree/main/Test_Cases/IEA-15-240-RWT-UMaineSemi/linearizations ? Or are you using updated linear models that you have generated yourself. 

If you are using your own linear model files - what is the difference? 

It looks to me like [this function](https://github.com/NREL/ROSCO/blob/77278f5d1cee95a830b22d79797b351481e87972/ROSCO_toolbox/linear/lin_util.py#L104) is calculating the stability margin as ~1 for all open-loop systems, which is very strange. If you are familiar with the process, it might be good to look at the nyquist plots of the openloop system itself to make sure they look reasonable. 

Thanks for your patience here - we are certainly still sorting out the best way to make these features accessible to everyone. There is certainly a lot of nuance to setting up this problem correctly, but hopefully it isn't too bad. 

Thanks,
Nikhar

**Comment by joeb8799** [comment]
Hello again Mr. Abbas,
I am very grateful for your answers. I will explain what I am trying to do.
I am trying to apply the robust scheduling method to https://github.com/NREL/ROSCO/tree/main/Test_Cases/IEA-15-240-RWT-UMaineSemi by generating my own linear models using OpenFAST so that I can understand how the procedure is done and then try to apply the robust scheduling method to any turbine model ( Automate it to any turbine model).

So first I changed some inputs in the OpenFAST input files, I looked to the example that you sent it earlier [iea15mw_baseline.tar.gz](https://github.com/NREL/ROSCO/files/8942525/iea15mw_baseline.tar.gz), and I changed some parameters in the .fst, Servodyn, Aerodyn15, Elastodyn and inflowfile. Then I get the linearization files for every wind speed of U_pc in the yaml file. Here is the link where you can see the inputs and outputs https://drive.google.com/drive/folders/19nVOvjfWsJGZkp1VPS-Tvtax14eAtefJ.
Then I tried to apply the robust scheduling to these linearization files and I had the error that you have already seen it.
I would like to mention that I had different linearization files I dont know why, is it possible to see the files and tell me where I did a mistake.

Thank you so much and have a good day.
Regards.

**Comment by Tommymaxprog** [comment]
I tried also to apply the Robust scheduling on the BAR_10MW, after generating the .lin files but I had the same issue that I had for the IEA-15-240-RWT-UMaineSemi.
Loading wind turbine data for NREL's ROSCO tuning and simulation processeses
-----------------------------------------------------------------------------
-----------------------------------------------------------------------------
   Tuning a reference wind turbine controller using NREL's ROSCO toolbox    
-----------------------------------------------------------------------------
Loading FAST model: BAR_10.fst 
Loading rotor performace data from text file: /home/g69581/ROSCO/Test_Cases/BAR_10/Cp_Ct_Cq.BAR_10.txt
Loading rotor performace data from text file: /home/g69581/ROSCO/Examples/../Test_Cases/BAR_10/Cp_Ct_Cq.BAR_10.txt
-----------------------------------------------------------------------------
Loading wind turbine data for NREL's ROSCO tuning and simulation processeses
-----------------------------------------------------------------------------
-----------------------------------------------------------------------------
   Tuning a reference wind turbine controller using NREL's ROSCO toolbox    
-----------------------------------------------------------------------------
Loading FAST model: BAR_10.fst 
Loading rotor performace data from text file: /home/g69581/ROSCO/Test_Cases/BAR_10/Cp_Ct_Cq.BAR_10.txt
Loading linearizations from: /home/g69581/ROSCO/Test_Cases/BAR_10/linearizations/BAR_10_07
Loading linearizations from: /home/g69581/ROSCO/Test_Cases/BAR_10/linearizations/BAR_10_09
loaded in parallel in 1.2891302108764648 seconds
Finding ROSCO tuning parameters for u = 12, sm = 0.1, omega_pc = [0.0, 0.3], k_float = 0.0
omega = 0.1, sm = 0.9999999999548844
omega = 0.1, sm = 0.9999999999548844
omega = 0.11000000000000001, sm = 0.9999999999548844
omega = 0.09, sm = 0.9999999999548844
omega = 0.11000000000000001, sm = 0.9999999999548844
omega = 0.09, sm = 0.9999999999548844
/home/g69581/anaconda3/envs/rosco-env/lib/python3.8/site-packages/openmdao/core/total_jac.py:1568: DerivativesWarning:Constraints or objectives [('r_sched.sm', inds=[0])] cannot be impacted by the design variables of the problem.
omega = 0.3, sm = 0.9999999999548844
omega = 0.31, sm = 0.9999999999548844
omega = 0.29, sm = 0.9999999999548844
Optimization terminated successfully    (Exit mode 0)
            Current function value: -0.3
            Iterations: 2
            Function evaluations: 2
            Gradient evaluations: 2
Optimization Complete
-----------------------------------
Finding ROSCO tuning parameters for u = 14, sm = 0.1, omega_pc = [0.0, 0.3], k_float = 0.0
omega = 0.1, sm = 0.9999999999509765
omega = 0.1, sm = 0.9999999999509765
omega = 0.11000000000000001, sm = 0.9999999999509765
omega = 0.09, sm = 0.9999999999509765
/home/g69581/anaconda3/envs/rosco-env/lib/python3.8/site-packages/openmdao/core/total_jac.py:1568: DerivativesWarning:Constraints or objectives [('r_sched.sm', inds=[0])] cannot be impacted by the design variables of the problem.
omega = 0.3, sm = 0.9999999999509765
omega = 0.31, sm = 0.9999999999509765
omega = 0.29, sm = 0.9999999999509765
Optimization terminated successfully    (Exit mode 0)
            Current function value: -0.3
            Iterations: 2
            Function evaluations: 2
            Gradient evaluations: 2
Optimization Complete
-----------------------------------
[0.3, 0.3] [1.2 1.2]

I really don't know why sm is equal to 1 for U_pc=12 and U_pc=14 (I chose U_pc=[12,14] in the yaml file) while it is given that sm=0.1 in the yaml file.
I hope that you can help me cause I don't know how to fix this issue by myself.
Thank you

**Comment by nikhar-abbas** [comment]
Hi,

It seems likely that, in both of these cases, something is wrong with either the linear model or the way in which the stability margin is being calculated. Have you looked at the Nyquist diagrams by any chance?

I just opened a pull request with some plotting functions here: https://github.com/NREL/ROSCO/pull/157. You might be able to emulate the code added to the end of example 12 to plot the Nyquist diagrams for your models as well.

Best,
Nikhar

**Comment by joeb8799** [comment]
Thank you for your response Mr.Abbas,
I will look at the Nyquist diagrams. But I have a question, how can I choose the number of inputs in the .lin files to be 10 and the number of continuous states to be 106 (similar to the linearization files present in https://github.com/NREL/ROSCO/tree/main/Test_Cases/IEA-15-240-RWT-UMaineSemi) Maybe the problem is in these parameters. I noticed that the number of inputs in my case is 160, which is 10 + 150 files related to the AeroDyn blades, how can I make it 10 and for the number of continuous states, I noticed that it depends on the degrees of freedom in the elastodyn file, so which degrees of freedom must be set true?
And one more question, the linearization files present in https://github.com/NREL/ROSCO/tree/main/Test_Cases/IEA-15-240-RWT-UMaineSemi are generated for an offshore wind turbine right? if it is true, what should be the value of CompMooring in the .fst file ? is it Map++?
I'm sorry to keep bothering you, thank you so much and have a good day sir.

**Comment by mohsamaz** [comment]
Hello everyone,
Could you please let me know where I can find detailed information or documentation about BAR_10?

Thank you in advance!
Best regards,

**Comment by ptrbortolotti** [comment]
BAR_10 was one of the designs developed during the BAR project (https://www.nrel.gov/wind/big-adaptive-rotor), which is now concluded. That design later evolved into the BAR-URC, which you can find here https://github.com/NREL/BAR_Designs and in the publications listed in the README
